strawberryfields/apps/train/cost.py

```diff
@@ -0,0 +1,117 @@
+# limitations under the License.
+r"""
+Submodule for computing gradients and evaluating cost functions with respect to GBS circuits
+
+In the context of stochastic optimization, cost functions are expressed as expectation values
+over the GBS distribution. Within the WAW parametrization, gradients of cost functions can be
+expressed as expectation values over the GBS distribution. This module contains methods for
+calculating these gradients and for using gradient-based methods to optimize GBS circuits. In the
+case of optimization with respect to a Kullback-Leibler divergence or log-likelihood cost
+function, gradients can be computed efficiently, leading to fast training.
+"""
+
+import numpy as np
+
+
+class KL:
+    r"""Kullback-Liebler divergence cost function.
+
+
+    In a standard unsupervised learning scenario, data are assumed to be sampled from an unknown
+    distribution and a common goal is to learn that distribution. Training of a model
+    distribution can be performed by minimizing the Kullback-Leibler (KL) divergence:
+    
+    .. math::
+
+        KL = -\frac{1}{T}\sum_S \log[P(S)]-\log(T),
+
+    where :math:`S` is an element of the data, :math:`P(S)` is the probability of observing that
+    element when sampling from the GBS distribution, and :math:`T` is the total number of elements
+    in the data. For the GBS distribution in the WAW parametrization, the gradient of the KL
+    divergence can be written as
+
+
+    .. math::
+
+        \partial_\theta KL(\theta) = - \sum_{k=1}^m(\langle n_k\rangle_{\text{data}}-
+        \langle n_k\rangle_{\text{GBS}})\partial_\theta w_k,
+
+    where :math:`\langle n_k\rangle` denotes the average photon numbers in mode *k*. This class
+    provides methods to compute gradients and evaluate the cost function.
+
+    **Example usage**
+
+    >>> embedding = embed.Exp(4)
+    >>> A = np.ones((4, 4))
+    >>> vgbs = param.VGBS(A, 3, embedding, threshold=True)
+    >>> params = np.array([0.05, 0.1, 0.02, 0.01])
+    >>> data = np.zeros((4, 4))
+    >>> kl = cost.KL(data, vgbs)
+    >>> kl.evaluate(params)
+    -0.2866830267216749
+    >>> kl.grad(params)
+    array([-0.52812574, -0.5201932 , -0.53282312, -0.53437824])
+
+    Args:
+        data (array): Array of samples representing the training data
+        vgbs: Variational GBS class
+
+    """
+
+    def __init__(self, data: np.ndarray, vgbs):
+        self.data = data
+        self.vgbs = vgbs
+        self.nr_samples = len(data)
+        self.nr_modes = len(data[0])
+        self.mean_n_data = np.mean(self.data, axis=0)
+
+    def grad(self, params: np.ndarray) -> np.ndarray:
+        r"""Calculates the gradient of the Kullback-Liebler cost function with respect to the
+        trainable parameters
+
+        **Example usage**
+
+        >>> kl.grad(params)
+        array([-0.52812574, -0.5201932 , -0.53282312, -0.53437824])
+
+        Args:
+            params (array): the trainable parameters :math:`\theta`
```

Comment:
Just to confirm, this should only work with real arrays right?
```suggestion
            params (array[float]): the trainable parameters :math:`\theta`
```

Reply:
Correct, the parameters must be real